### DISCUSSION TITLE
A bit of a weird `dispose` pattern - Concerns on leaky resources?

### DISCUSSION THREAD

**Discussion by TomTheFurry** [amerkoleci/Vortice.Vulkan#52] [Q&A] [answered]
I have been getting an (unrelated) issue with glsl include sub-includes sometimes seemingly (probably?) accessing freed file data. So, when I am debugging that, I noticed in here, if the `Result` object isn't explicitly disposed, (i.e. instead, letting c# GC handle it), the `~Result()` finalizer actually doesn't do anything? Cause the native `Result` object is apparently only freed if you call the explicit `Dispose()`?

I have also seen this appears in many other places, with (I think) native resources being conditionally freed behind that bool flag. So I am really not sure if this is a bug or if I am missing something?

Code:
https://github.com/amerkoleci/Vortice.Vulkan/blob/984b67941e2410816c5c8a812f20c7d5f4018ab5/src/Vortice.ShaderCompiler/Result.cs#L37-L57

**Comment by amerkoleci** [comment]
Can you try to follow the IDisposable pattern using keyword and see if it causes issue?

Take look here:
https://github.com/amerkoleci/Vortice.Vulkan/blob/main/tests/Vortice.ShaderCompiler.Test/CompileTests.cs#L19

**Reply by TomTheFurry** [reply]
Yeah I know about that. But it's just that the code seems to indicate it wants to support GC based self disposal, but actually doesn't really.

**Reply by TomTheFurry** [reply]
So I guess question is whether user of the API is expected to *always* call `Dispose()` for those api objects, or if c# GC can handle it themselves.

**Reply by TomTheFurry** [reply]
(Btw not related much, but I did try and see adding dispose help fixes my `#include` corruption issue. Well, it now insteads straight up throws `AccessViolationException` on the compile call. So.... not going great. Gonna spend more time debugging and see what's happening)

**Reply by amerkoleci** [reply]
Share some code so I can gladly help

**Reply by TomTheFurry** [reply]
Trying to zoom in the scope atm, and extracting a standalone repro case. (I was doing some company code, so, yeah.)

**Reply by TomTheFurry** [reply]
![image](https://github.com/user-attachments/assets/0817401a-bade-4388-9fd1-80c847bb12e5)
And... how the heck did a `#line` directive even happen!? I swear nowhere in any shader file contains any `#line` directives!

*Plz don't tell me it's a deeper shaderc issue...*

**Reply by TomTheFurry** [reply]
![image](https://github.com/user-attachments/assets/dedfe6d3-ff79-431e-933c-c43377f7df1c)
Okay yeah not sure what's going on. It just sometimes randomly fails on a  `#include`, and retrying again and again then sometimes works. Hmm...

**Reply by amerkoleci** [reply]
Just released version `https://www.nuget.org/packages/Vortice.ShaderCompiler/1.8.0` which improves all this stuff, can you test?

**Comment by TomTheFurry** [comment, marked as answer]
Update, can confirm that the latest (ShaderCompiler v1.8.0) did fix above mentioned issues!

Btw, it might be a good idea to add comments on `Compiler` class that it **REQUIRES** explicit call to `Dispose()`.

### Side-note:
My current guess on why the shaderc is reading corrupted data on `#include` is that the old version of passing included file strings into native environment seems to pass incorrect 'length' param compared to the raw utf-8 c-string it passes:

https://github.com/amerkoleci/Vortice.Vulkan/blob/f75763a56960b9ad9c2b0aa26552fc637d0d3933/src/Vortice.ShaderCompiler/Includer.cs#L75-L76

In above, it is passing down the utf-16 c#-style string `Length` as the utf-8 string's 'length' parameter, meaning a mismatch of what it say the string length is vs what the actual string length is.

Now, how, or even *if*, those mismatches lead to the memory corruption, I have no idea. This really is just a guess. *It prob was causing problems. But is that the specific problem I was encountering? Who knows.*

I can confirm however that the new reworked code indeed does this properly, and that the issues on corrupted memory no longer (seems to) happen. Thanks @amerkoleci a lot!
